packages/babel-plugin-proposal-decorators/CONTRIB.md

```diff
@@ -0,0 +1,374 @@
+decorators. This is important because we must apply the decorators _after_ the
+class prototype has been fully setup, but _before_ static fields are run, since
+static fields should only see the decorated version of the class.
+
+We apply the decorators to class elements and the class itself, and the
+application returns an array of values that are used to populate all of the
+local variables we defined earlier. The array's order is fully deterministic, so
+we can assign the values based on an index we can calculate ahead of time.
+
+We'll come back to `applyDecs` in a bit to dig into what its format is exactly,
+but now let's dig into the new definitions of our class elements.
+
+```js
+  a = (initInstance(this), initA(this, 123));
+```
+
+Alright, so previously this was a simple class field. Since it's the first field
+on the class, we've updated it to immediately call `initInstance` in its
+initializer. This calls any initializers added with `addInitializer` for all of
+the per-class values (methods and accessors), which should all be setup on the
+instance before class fields are assigned. Then, it calls `initA` to get the
+initial value of the field, which allows initializers returned from the
+decorator to intercept and decorate it. It's important that the initial value
+is used/defined _within_ the class body, because initializers can now refer to
+private class fields, e.g. `a = this.#b` is a valid field initializer and would
+become `a = initA(this, this.#b)`, which would also be valid. We cannot
+extract initializer code, or any other code, from the class body because of
+this.
+
+Overall, this decoration is pretty straightforward other than the fact that we
+have to reference `initA` externally.
+
+```js
+  static #b(...args) {
+    callB(this, args);
+  }
+  static x() {
+    console.log('foo');
+  }
+```
+
+Next up, we have a private static class method `#b`. This one is a bit more
+complex, as our definition has been broken out into 2 parts:
+
+1. `static #b`: This is the method itself, which being a private method we
+  cannot overwrite with `defineProperty`. We also can't convert it into a
+  private field because that would change its semantics (would make it
+  writable). So, we instead have it proxy to the locally scoped `callB`
+  variable, which will be populated with the fully decorated method.
+2. `static x`: This contains the _code_ of the original method. Once again, this
+  code cannot be removed from the class body because it may reference private
+  identifiers. However, we have moved the code to a _public_ method, which means
+  we can now read its value using `Object.getOwnPropertyDescriptor`. Decorators
+  use this to get the initial implementation of the method, which can then be
+  wrapped with decorator code/logic. They then `delete` this temporary property,
+  which is necessary because no additional elements should be added to a class
+  definition.
```

Comment:
Is this deleted property somewhere observable? i.e. can decorators access `_Class` before it's deleted?

We might consider storing it as a private method (which then doesn't need to be deleted), and pass an additional array of "private accessors" to `_applyDecorators`, like
```js
applyDecs(
    this,
    elementDecs,
    classDecs,
    [o => o.#original_b]
);
```

If we end up using the "private method to fields" design suggested in the above comment, then we only need to do
```js
static #b = function () {
  console.log('foo');
}
```
and then pass it like
```js
applyDecs(
    this,
    elementDecs,
    classDecs,
    [this.#b]
);
```

Reply:
So I came up with this before I realized that we could use static blocks (and in fact needed to in order to match the spec), and I just tested it out, we can actually refer to private names in static blocks. So we can fully extract private methods to the static block and avoid the delete in these cases 😄 I'll think about the best way to do this and update the code of `applyDecs` accordingly, no need to add private class fields.